plugins/tech-insights-backend/src/types.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Config } from '@backstage/config';
+import { DateTime } from 'luxon';
+import {
+  DynamicFactCallback,
+  Event,
+  FactOptions,
+  TopLevelCondition,
+} from 'json-rules-engine';
+import { PluginEndpointDiscovery } from '@backstage/backend-common';
+
+export type TechInsightFact = {
+  ref: string;
+  entity: {
+    namespace: string;
+    kind: string;
+    name: string;
+  };
+  facts: Record<string, number | string | boolean | DateTime>;
+};
+
+type FactValueDefinitions = {
+  [key: string]: {
+    type: 'integer' | 'float' | 'string' | 'boolean' | 'datetime';
+    description?: string;
+    since?: string;
+  };
+};
+export type FactSchema = {
+  version: string;
+  schema: FactValueDefinitions;
+};
+
+export type FactRetrieverContext = {
+  config: Config;
+  discovery: PluginEndpointDiscovery;
+  // ...others
+};
+
+export interface FactRetriever {
```

Comment:
So one assumption I notice is that a FactRetriever seems to operate on all entities. Now it's possible that the retriever can decide which entities it's applicable to and only process those. I'm wondering though if there's any possibility that users may want to configure this and if we need it in the API here? Or perhaps we need a means of user supplied options generally?

Reply:
yeah, that's a bit tricky question to think about. Now a fact retriever essentially can do what ever it wants. It can get all entities it wants via the the passed in config/discovery API or do anything else. Something could be that we find a way to indicate that only a subset of entities are managed by this retriever and then those entities would be passed in to the handler as part of the context. It might be a bit too open now.

Reply:
User supplied options is a good shout and should be added in def

Reply:
Moved cadence outside of the interface. Opting to keep the context passed into the handler as 'config' & 'discovery' for now.